lib/auth/touchid/credentials.m

```diff
@@ -0,0 +1,110 @@
+
+#include "credentials.h"
+
+#import <CoreFoundation/CoreFoundation.h>
+#import <Foundation/Foundation.h>
+#import <Security/Security.h>
+
+#include <stdlib.h>
+
+#include "common.h"
+
+BOOL matchesLabelFilter(LabelFilterKind kind, NSString *filter,
+                        NSString *label) {
+  switch (kind) {
+  case LABEL_EXACT:
+    return [label isEqualToString:filter];
+  case LABEL_PREFIX:
+    return [label hasPrefix:filter];
+  }
+  return NO;
+}
+
+int FindCredentials(LabelFilter filter, CredentialInfo **infosOut) {
+  NSDictionary *query = @{
+    (id)kSecClass : (id)kSecClassKey,
+    (id)kSecAttrKeyType : (id)kSecAttrKeyTypeECSECPrimeRandom,
+    (id)kSecMatchLimit : (id)kSecMatchLimitAll,
+    (id)kSecReturnRef : @YES,
+    (id)kSecReturnAttributes : @YES,
+  };
+  CFArrayRef items = NULL;
+  OSStatus status =
+      SecItemCopyMatching((CFDictionaryRef)query, (CFTypeRef *)&items);
+  switch (status) {
+  case errSecSuccess:
+    break; // continue below
+  case errSecItemNotFound:
+    return 0; // aka no items found
+  default:
+    // Not possible afaik, but let's make sure we keep up the method contract.
+    if (status >= 0) {
+      status = status * -1;
+    }
+    return status;
+  }
+
+  NSString *nsFilter = [NSString stringWithUTF8String:filter.value];
+
+  CFIndex count = CFArrayGetCount(items);
+  *infosOut = calloc(count, sizeof(CredentialInfo));
+  int infosLen = 0;
+  for (CFIndex i = 0; i < count; i++) {
+    CFDictionaryRef attrs = CFArrayGetValueAtIndex(items, i);
+
+    CFStringRef label = CFDictionaryGetValue(attrs, kSecAttrLabel);
+    NSString *nsLabel = (__bridge NSString *)label;
+    if (!matchesLabelFilter(filter.kind, nsFilter, nsLabel)) {
+      continue;
+    }
+
+    CFDataRef appTag = CFDictionaryGetValue(attrs, kSecAttrApplicationTag);
+    NSString *nsAppTag =
+        [[NSString alloc] initWithData:(__bridge NSData *)appTag
+                              encoding:NSUTF8StringEncoding];
+
+    CFDataRef appLabel = CFDictionaryGetValue(attrs, kSecAttrApplicationLabel);
+    NSString *nsAppLabel =
+        [[NSString alloc] initWithData:(__bridge NSData *)appLabel
+                              encoding:NSUTF8StringEncoding];
+
+    // Copy public key representation.
+    SecKeyRef privKey = (SecKeyRef)CFDictionaryGetValue(attrs, kSecValueRef);
+    SecKeyRef pubKey = SecKeyCopyPublicKey(privKey);
+    char *pubKeyB64 = NULL;
+    if (pubKey) {
+      CFDataRef pubKeyRep =
+          SecKeyCopyExternalRepresentation(pubKey, NULL /*error*/);
+      if (pubKeyRep) {
+        NSData *pubKeyData = CFBridgingRelease(pubKeyRep);
+        pubKeyB64 = CopyNSString([pubKeyData base64EncodedStringWithOptions:0]);
+      }
+      CFRelease(pubKey);
+    }
+
+    (*infosOut + infosLen)->label = CopyNSString(nsLabel);
+    (*infosOut + infosLen)->app_label = CopyNSString(nsAppLabel);
+    (*infosOut + infosLen)->app_tag = CopyNSString(nsAppTag);
+    (*infosOut + infosLen)->pub_key_b64 = pubKeyB64;
```

Comment:
Why not `infosOut[infosLen]->label = ...`?

Reply:
These are not equivalent. We could do `(*infosOut)[infosLen].label = ...`, but I think I'd rather keep it as-is.